README

==== OVERVIEW ====

    The Regexp::Common::debian is collection of REs for various strings
    found in the Debian Porject <http://debian.org>.  It's no way intended
    to be a validation tool.

    R::C::d needs perl ``v5.10.0'' or later because:

:   *   4 patterns (29%) make use of decent perls (one doesn't make it
        through with anything older);

:   *   It's time to move, ``v5.10.0'' is six year old and three years
        no-support;

:   *   ``lenny'' is four year old and a year no-support;


==== INSTALL ====

    The R::C::d builds with Module::Build.

        $ perl Build.PL
        $ perl Build
        $ perl Build test
        $ perl Build install

    Since we're about strings we need a lots of strings to test against
    (Test::More, unspecified version).  To access them easily (it's all
    about reuse, not implemented yet) I need an apropriate storage. 
    Accidentally it's YAML::Tiny (unspecified version).

    ``v0.2.1'' Reading reports of cpantesters I've got to conclusion that
    YAML::Tiny isn't popular.  (``v0.2.13'' Wandering through errors of
    ``v0.2.12'' I should say it totally is.) And avoiding installing (or
    unability to install (there could be reasons)) build requirements isn't
    that uncommon.  Although I experience a strong belief that some YAML
    reader happens to be installed anyway.  And still I can't find a way to
    specify that *%build_requires* one of but all known to me YAML reader. 
    So here is a dirty trick.  t::TestSuite attempts to require() one of
    known (to me, see below) YAML reader.  Then (upon initial ``perl
    Build.PL'') t::TestSuite is asked what it has found (if nothing then
    cosmetic ``Compilation failed in require'' message will be seen).  And
    one what has been found will be added to *%build_requires*; If nothing
    then fair YAML::Tiny will be added.  (I think it's fair because
    YAML::Tiny is pure-Perl, small footprint, and no dependencies.)

    (note) I'm talking about "known to me YAML readers" because I've found
    out that different YAML readers treat source differently.  So I attempt
    to keep t/*.yaml files semantically equal and sintactically correct. 
    Hopefully there're no differences among versions in wild.

    ``v0.2.13'' (Actually, this feature has been here for years.) Any
    supported YAML reader can be enforced with $ENV{RCD_YAML_ENGINE} magic
    (in spite of any build-time choice):

        RCD_YAML_ENGINE=syck ./Build test

    Readers are assigned by nicks.  Here they are:

:   *   ``syck'' -- YAML::Syck.

:   *   ``xs'' -- YAML::XS.

:   *   ``tiny'' -- YAML::Tiny.

:   *   ``old'' -- YAML.

:   *   ``data'' -- Data::YAML::Reader 'does not support multi-line quoted
        scalars', 'YAML document header not found' -- unsupported, so far.

    ``v0.2.2'' Various (all, except t/preferences.t and t/sourceslist.t)
    test-units know a magic command '$ENV{RCD_ASK_DEBIAN}'.  Apply it this
    way (enabling all possible external inquiries):

        RCD_ASK_DEBIAN=all ./Build test

    or this (separate keys with any non-word):

        RCD_ASK_DEBIAN=binary,architecture ./Build test

    When applied a test-unit would ask Debian's commands or inspect Debian
    specific files for information the test-unit is interested in.  For
    obvious reasons that magic will fail on non-Debian system; So don't. 
    Although if used correctly that could warn of strange ('not known
    before') compatibility problems.  Details:

:   architecture of t/architecture.t
        This asks ``dpkg-architecture -L'' for list of known architectures
        (per Section 11.1 of debian-policy).  That wouldn't find
        architectures dropped (had that happen ever?) but omissions won't
        stay unnoted anymore.

:   binary of t/archive.misc.t
        ``v0.2.3'' Inspects all records in /var/lib/apt/lists/*_Packages,
        extracts *Filename:* entries and matches all of them against
        ``m/^$RE{debian}{archive}{binary}$/''.  All (if any) failure will be
        reported at the end.

:   changelog of t/changelog.t
        ``v0.2.8'' That will inspect /usr/share/doc/*/changelog.Debian
        files.  To do a complete scan it would take loads of time (really). 
        You should understand, that's not enough to just run through
        changelogs.  It has to be verified that none entry is skipped.  The
        only reliable (for sake of interface, and, trivially, presence)
        source of verification is ``dpkg-parsechangelog''.  And here's the
        fork-mare.  ``perl'' forks ``shell'', then ``perl'', then ``perl''
        again.  There seems to be fork of ``tail'' too.  And that for each
        entry.  (Not to count ``gunzip'' to decompress the changelog.)
        ``loadavg'' climbs over 1.50..2.00 You've got the picture.  Although
        that's where choice begins.

        ``v0.2.12'' That happens that ``urgency=high'', probably when it's
        that really high, is expressed in blocks (like this:
        ``urgency=HIGH'').  $RE{d}{changelog} is case-keeping, and then
        dpkg-parsechangelog(1) strikes back and lowercase.  From now on such
        manipulations won't fail a particular entry.

:       *changelog*
            ``v0.2.9'' That defaults to ``changelog=5''.  See below.

:       *changelog=package*
            Only one changelog will be checked.  The one that ``eq''s.  The
            package name is picked from directory name.

:       *changelog=a*
            Only those changelogs will be checked that ``m/^a/''.

:       *changelog=5*
            ``v0.2.9'' That will check all changelogs, although it will look
            no more than requested number of entries deep.  (``v0.2.9'':
            ~15min for ~1200 changelogs; ``v0.2.12'': ~30min for ~1300
            changelogs and ~6000 subchecks.) And that has a perfect sense. 
            Do you know that ``cron'' once changed it's name to ``Cron''
            (beware leading block) (cron_3.0pl1-46)? C'mon, it has happened
            12 (tweleve) years ago! (And you know what? That default is
            pretty fair (liblog-log4perl-perl_1.16-1).  Probably it should
            look for time passed but entry number.)

:       *changelog=-5*
            ``v0.2.9'' That's different.  It will check as many entries as
            possible (there are changelogs what $RE{d}{changelog} finds out
            more entries than dpkg-parsechangelog (``dpkg_1.2.13'' vs
            ``dpkg_0.93.79''), but if the offending record is more than that
            far from top then it's reported and otherwise ignored. 
            (``v0.2.9'': ~3h for ~1200 packages, ~45000 successful
            subchecks; ``v0.2.12'': ~5¼h for ~1300 changelogs, ~63000
            successful subchecks.)

:       *changelog=_5*
            ``v0.2.12'' That's a mix of *changelog=5* and *changelog=-5*
            (thanks to irda-utils_0.9.18-8.1 and mime-support_3.49-1).  It
            goes no more than configured entries deep and ignores (and
            reports) any errors.

:       *changelog=0*
            (bug) ``v0.2.9'' That will check all changelogs, check all
            possible entries and BAIL_OUT off first failure.  Shortly --
            don't.  You're warned.  (Although, do it.  t/changelog.t will
            give up pretty soon.)

        To slightly sweeten all that, t/changelog.t attempts to filter
        duplicates.  And it BAIL_OUTs upon first failure.

:   package of t/package.t
        ``v0.2.10'' Nothing special.  Output of ``dpkg-query -f
        '${Package}\n' -W'' is matched against m/^$RE{debian}{package}$/. 
        Probably should parse *_Packagees.

:   source of t/archive.source.t
        ``v0.2.3'' Inspects all records in /var/lib/apt/lists/*_Sources,
        extracts *Files:* entries, then collects trailing filenames.  They
        are matched against ``m/^$RE{debian}{archive}{source_1_0}$/'',
        ``m/^$RE{debian}{archive}{patch_1_0}$/'',
        ``m/^$RE{debian}{archive}{source_3_0_native}$/'',
        ``m/^$RE{debian}{archive}{source_3_0_quilt}$/'',
        ``m/^$RE{debian}{archive}{patch_3_0_quilt}$/'', and
        ``m/^$RE{debian}{archive}{dsc}$/'' (in fact ``||'').  If none
        matches then it will be reported at the end. 
        ``m/$RE{debian}{archive}{changes}/'' is missing here because there
        is no source of such on no-build system.

:   version of t/version.t
        ``v0.2.10'' Again nothing special.  Output of ``dpkg-query -f
        '${Version}\n' -W'' is matched against m/^$RE{debian}{version}$/. 
        Probably should parse *_Packages too.

    If any test string fails I need to know what and how.  To provide that
    info I've picked Test::Differences (maybe there's other option I'm not
    aware of?) (I'm, Test::Deep).  ``v0.60'' of T::D closes
    [38320@rt.cpan.org] and [41241@rt.cpan.org].


==== AVAILABILITY ====

    Distribution -- <http://search.cpan.org/dist/Regexp-Common-debian/>


==== BUGS ====

    Please report here --
    <http://rt.cpan.org/Public/Dist/Display.html?Name=Regexp-Common-debian>


==== COPYRIGHT AND LICENSING ====

:   *   Copyright 2008--2010, 2014 Eric Pozharski <whynot@cpan.org>

:   *   AS-IS, NO-WARRANTY, HOPE-TO-BE-USEFUL

:   *   GNU Lesser General Public License v3
        <http://www.gnu.org/licenses/lgpl.txt>



Changes
  v0.2.13 20140205

      % TS' data-sets made readable by all supported YAML readers
          Breakage has been here for years.

      % Adding YAML::XS reader
      % Data::YAML left unsupported
      % Adding *dynamic_config* to Build.PL
          Breakage has been here for years.

  v0.2.12 20140117

      ~ $RE{debian}{architecture} (+26 -4 ~3 =42)
          Lame attempt to be RL-correct has been removed -- any *os* and
          *cpu* will fit; "any" can take over *os* and/or *cpu*.  Lists
          of *os*es and *cpu*s have been updated.

      ~ $RE{debian}{archive}{changes} (=41)
          New *os* and *cpu* lists have been canibalized.  Using
          "v5.10.0" features and requiring it.

      ~ $RE{debian}{archive}{source_3_0_native} (+3 =66)
      ~ $RE{debian}{archive}{source_3_0_quilt} (+1 =111)
      ~ $RE{debian}{archive}{patch_3_0_quilt} (+1 =72)
          "xz" suffix has been added.

      ~ $RE{debian}{archive}{binary} (=59)
          New *os* and *cpu* lists have been canibalized.

      ~ $RE{debian}{preferences} (+4 =74)
      ~ $RE{debian}{changelog} (=145)
      ~ $RE{debian}{sourceslist} (=26)
          Using "v5.10.0" features and requiring it.

      - workaround for perl pre v5.10.0 for $RE{d}{version} (-45 =0)
          Workaround has been retired.

      % Test-suite support testing live-data from various sources
          Filtering duplicates.  Adding some work-arounds about RL.

      % Documentation fixes
      = XXX scores 7
      = TODO scores 16
      = FIXME scores 17

  v0.2.11 20100808

      ~ $RE{debian}{version} (=45)
          Support for versions with no leading number.

      + $RE{debian}{archive}{source_3_0_native} (63)
      + $RE{debian}{archive}{source_3_0_quilt} (110)
      + $RE{debian}{archive}{patch_3_0_quilt} (71)
          Support for filenames of "Format: 3.0".

      + $RE{debian}{archive}{source_1_0} (+16 =59)
      - $RE{debian}{archive}{source}
          Follow up rename.  Support for versions with no leading
          number.

      + $RE{debian}{archive}{patch_1_0} (+13 =65)
      - $RE{debian}{archive}{patch}
          Follow up rename.  Support for versions with no leading
          number.

      ~ $RE{debian}{archive}{binary} (+9 =59)
      ~ $RE{debian}{archive}{changes} (+7 =41)
      ~ $RE{debian}{archive}{dsc} (+12 =62)
          Support for versions with no leading number.

      ~ $RE{debian}{changelog} (+30 =145)
          Support for versions with no leading number.  Various fixes to
          tolerate variations in emtpy lines, leading and separating
          spaces.  Support for urgency comments.

      ~ $RE{debian}{architecture} (+4 =20)
          List of acceptable architectures has been updated.

      ~ $RE{debian}{preferences} (-4 =70)
      ~ $RE{debian}{sourceslist} (-4 =26)
          Cosmetic changes.

      % Test-suite support testing live-data from various sources
      % Documentation fixes
      = XXX scores 8
      = TODO scores 12
      = FIXME scores 13

  v0.1.5 20090311

      ~ $RE{debian}{changelog} (+34 =115)
      % Documentation fixes.

  v0.1.4 20090209

      + $RE{debian}{sourceslist} (30)
      + $RE{debian}{preferences} (74)
      + $RE{debian}{changelog} (81)
      % TestSuite.pm does one less string eval.
      % Documentation upgraded.
      % Various changes and updates to satisfy CPAN.

  v0.0.11 20090103

      % $RE{debian}{archive}{changes} was hardly broken.
          Copy-paste issue, fixed.

      % Requiring perl v5.6.0
          Since our is in use.

      % Cosmetic documentation updates-n-fixes.
      ~ $RE{debian}{archive}{changes} (-16 =34)

  v0.0.10 20081231

      + $RE{debian}{package} (69)
      + $RE{debian}{version} (45)
      + workaround for perls pre v5.10.0 for $RE{d}{version} (45)
      + $RE{debian}{architecture} (16)
      + $RE{debian}{archive}{binary} (50)
      + $RE{debian}{archive}{source} (43)
      + $RE{debian}{archive}{patch} (52)
      + $RE{debian}{archive}{dsc} (50)
      + $RE{debian}{archive}{changes} (50)
      = XXX scores 8
      = TODO scores 5
      = FIXME scores 8



t/TestSuite.pm
# $Id: TestSuite.pm 394 2010-08-07 15:10:03Z whynot $

package t::TestSuite;

use strict;
use warnings;
use version 0.50; our $VERSION = qv q|0.2.4|;
use base qw| Exporter |;
use lib   q|./blib/lib|;

use Cwd;
use Test::Differences;
use Data::Dumper;
use Module::Build;

our @EXPORT_OK = qw| RCD_process_patterns |;

$ENV{PERL5LIB} = getcwd . q(/blib/lib);

# FIXME: B<&Module::Build::runtime_params> apeared in v0.28
our $Verbose = eval { Module::Build->current->runtime_params(q|verbose|); };

our $Y_Choice;
foreach my $y_eng
( [qw| syck YAML::Syck |],
  [qw| old  YAML       |],
  [qw| tiny YAML::Tiny |]) {
    $ENV{RCD_YAML_ENGINE} && $y_eng->[0] ne $ENV{RCD_YAML_ENGINE}    and next;
    eval qq|require $y_eng->[1]|                                      or next;
    $Y_Choice = $y_eng;
    last                    }
$Y_Choice                     or die q|none known YAML reader has been found|;

sub RCD_show_y_choice ( ) {                       print qq|$Y_Choice->[1]\n| }

sub RCD_load_patterns ( )               {
    my $fn = (caller)[1];
    $fn =~ s{\.t$}{.yaml};
    if( $Y_Choice->[0] eq q|tiny| )    {
        my $yaml = YAML::Tiny->read($fn);
        return %{$yaml->[0]}            }
    elsif( $Y_Choice->[0] eq q|old| )  {
        my( $yaml, $buf );
        open my $fh, q|<|, $fn;
        read $fh, $buf, -s $fh;
        $yaml = YAML::Load( $buf );
        return %$yaml                   }
    elsif( $Y_Choice->[0] eq q|syck| ) {
        my $yaml = YAML::Syck::LoadFile( $fn );
        return %$yaml                   }}

sub RCD_save_patterns ( $\% )             {
    my($fn, $data) = ( @_ );
    if( $Y_Choice->[0] eq q|tiny| )      {
        my $yaml = YAML::Tiny->new;
        $yaml->[0] = $data;
        $yaml->write($fn)                 }
    elsif( $Y_Choice->[0] eq q|old| )    {
        open my $fh, q|>|, $fn;
        print $fh YAML::Dump( $data )     }
    elsif( $Y_Choice->[0] eq q|syck| )   {
        YAML::Syck::DumpFile( $fn, $data )}}

sub RCD_process_patterns ( % ) {
    my %args = ( @_ );

    foreach my $ptn (@{$args{patterns}})                               {
        my @in =  ( @$ptn );
        my @out = (
          $ptn->[0],
          ($ptn->[0] =~ $args{re_m} ? '+' : '-'),
          $ptn->[0] =~ $args{re_g} );
        my $dump = Data::Dumper
          ->new([ $ptn->[0] ])->Terse(1)->Useqq(1)->Indent(0)->Dump;
        $dump =~ s{^"(.+)"$}{$1};
        eq_or_diff_data \@out, \@in, sprintf q|%s %s|, $ptn->[1], $dump }

    Test::More::diag(
      sprintf q|processed patterns (%s): %i|,
        ( caller 1 )[3] || ( caller )[1], scalar @{$args{patterns}})
        if $Verbose             }

sub RCD_trial_patterns ( % ) {
    my %args = @_;
    my @rc;

    push @rc, [ $_, (m[$args{re_m}] ? '+' : '-'), m[$args{re_g}] ]
      foreach @{$args{patterns}};
    @rc                       }

sub RCD_count_patterns ( )         {
    opendir my $dh, q|t|;
    while( my $fn = readdir $dh ) {
        index( $fn, '.' ) && $fn =~ m{.yaml}                          or next;
        my %patterns;
        if( $Y_Choice->[0] eq q|tiny| )    {
            my $yaml = YAML::Tiny->read(qq|t/$fn|);
            %patterns = %{$yaml->[0]}       }
        elsif( $Y_Choice->[0] eq q|old| )  {
            my( $yaml, $buf );
            open my $fh, q|<|, qq|t/$fn|;
            read $fh, $buf, -s $fh;
            $yaml = YAML::Load( $buf );
            %patterns = %$yaml              }
        elsif( $Y_Choice->[0] eq q|syck| ) {
            my $yaml = YAML::Syck::LoadFile( qq|t/$fn| );
            %patterns = %$yaml              }
        printf qq|%23s => % 4i\n|, $_, scalar @{$patterns{$_}}
          foreach keys %patterns   }}

1;


TestSuite.pm
# $Id: TestSuite.pm 9 2008-12-12 03:37:21Z whyn0t $

package TestSuite;

use strict;
use warnings;
use version 0.50;

use base qw(Exporter);
use vars qw(@EXPORT_OK);
use Cwd;
use YAML::Tiny;
use Test::Differences;

our $VERSION = qv q|0.0.2|;

use lib q(./blib/lib);

@EXPORT_OK = qw| RCD_process_patterns |;

$ENV{PERL5LIB} = getcwd . q(/blib/lib);

sub RCD_do_units (\@@)       {
    my($units) = shift @_;
    unless(@_)            {
        $_->()
          foreach @$units; }
    else                  {
        eval qq|main::RCD_$_| or
          Test::More::diag($@)
          foreach @_;      }; };

sub RCD_load_patterns () {
    my $fn = (caller)[1];
    $fn =~ s{\.t$}{.yaml};
    my $yaml = YAML::Tiny->read($fn);
    return %{$yaml->[0]}; };

sub RCD_save_patterns ($\%) {
    my($fn, $data) = ( @_ );
    my $yaml = YAML::Tiny->new;
    $yaml->[0] = $data;
    $yaml->write($fn);       };

sub RCD_process_patterns (%)     {
    my %args = ( @_ );
    my(@in, @out);

    foreach my $ptn (@{$args{patterns}})         {
        my @in =  ( @$ptn );
        my @out = (
          $ptn->[0],
          ($ptn->[0] =~ $args{re_m} ? '+' : '-'),
          $ptn->[0] =~ $args{re_g} );
        eq_or_diff_data
          \@out,
          \@in,
          sprintf q|%s %s|, $ptn->[1], $ptn->[0]; };

    Test::More::diag(
      sprintf q|processed patterns (%s): %i|,
      (caller 1)[3],
      scalar @{$args{patterns}}); };

1;


Build.PL
#!/usr/bin/perl
# $Id: Build.PL 493 2014-02-05 19:00:07Z whynot $

use strict;
use warnings;
package main;

use version 0.50; our $VERSION = qv q|0.2.14|;

use Module::Build 0.38;

=head1 OVERVIEW

The B<Regexp::Common::debian> is collection of REs for various strings found in
the Debian Porject <http://debian.org>.
It's no way intended to be a validation tool.

B<R::C::d> needs perl C<v5.10.0> or later because:

=over

=item *

4 patterns (29%) make use of decent perls
(one doesn't make it through with anything older);

=item *

It's time to move, C<v5.10.0> is six year old and three years no-support;

=item *

C<lenny> is four year old and a year no-support;

=back

=head1 INSTALL

The B<R::C::d> builds with B<Module::Build>.

    $ perl Build.PL
    $ perl Build
    $ perl Build test
    $ perl Build install

Since we're about strings we need a lots of strings to test against
(B<Test::More>, unspecified version).
To access them easily (it's all about reuse, not implemented yet) I need an
apropriate storage.
Accidentally it's B<YAML::Tiny> (unspecified version).

C<v0.2.1>
Reading reports of cpantesters I've got to conclusion that B<YAML::Tiny> isn't
popular.
(C<v0.2.13>
Wandering through errors of C<v0.2.12> I should say it totally is.)
And avoiding installing (or unability to install (there could be reasons))
build requirements isn't that uncommon.
Although I experience a strong belief that some YAML reader happens to be
installed anyway.
And still I can't find a way to specify that I<%build_requires> B<one of> but
B<all> known to me YAML reader.
So here is a dirty trick.
B<t::TestSuite> attempts to B<require()> one of known (to me, see below) YAML
reader.
Then (upon initial C<perl Build.PL>) B<t::TestSuite> is asked what it has found
(if nothing then cosmetic C<Compilation failed in require> message will be
seen).
And one what has been found will be added to I<%build_requires>;
If nothing then fair B<YAML::Tiny> will be added.
(I think it's fair because B<YAML::Tiny> is pure-Perl, small footprint, and no
dependencies.)

B<(note)>
I'm talking about "known to me YAML readers" because I've found out that
different YAML readers treat source differently.
So I attempt to keep F<t/*.yaml> files semantically equal and sintactically
correct.
Hopefully there're no differences among versions in wild.

C<v0.2.13>
(Actually, this feature has been here for years.)
Any supported YAML reader can be enforced with C<$ENV{RCD_YAML_ENGINE}> magic
(in spite of any build-time choice):

    RCD_YAML_ENGINE=syck ./Build test

Readers are assigned by nicks.
Here they are:

=over

=item *

C<syck> -- B<YAML::Syck>.

=item *

C<xs> -- B<YAML::XS>.

=item *

C<tiny> -- B<YAML::Tiny>.

=item *

C<old> -- B<YAML>.

=item *

C<data> -- B<Data::YAML::Reader> 'does not support multi-line quoted scalars',
'YAML document header not found' -- unsupported, so far.

=back

C<v0.2.2>
Various (all, except F<t/preferences.t> and F<t/sourceslist.t>)
test-units know a magic command
'C<$ENV{RCD_ASK_DEBIAN}>'.
Apply it this way (enabling all possible external inquiries):

    RCD_ASK_DEBIAN=all ./Build test

or this (separate keys with any non-word):

    RCD_ASK_DEBIAN=binary,architecture ./Build test

When applied a test-unit would ask Debian's commands or inspect Debian specific
files for information the test-unit is interested in.
For obvious reasons that magic will fail on non-Debian system;
So don't.
Although if used correctly that could warn of strange ('not known before')
compatibility problems.
Details:

=over

=item B<architecture> of F<t/architecture.t>

This asks C<dpkg-architecture -L> for list of known architectures
(per Section 11.1 of debian-policy).
That wouldn't find architectures dropped (had that happen ever?)
but omissions won't stay unnoted anymore.

=item B<binary> of F<t/archive.misc.t>

C<v0.2.3>
Inspects all records in F</var/lib/apt/lists/*_Packages>, extracts I<Filename:>
entries and matches all of them against
C<m/^$RE{debian}{archive}{binary}$/>.
All (if any) failure will be reported at the end.

=item B<changelog> of F<t/changelog.t>

C<v0.2.8>
That will inspect F</usr/share/doc/*/changelog.Debian> files.
To do a complete scan it would take B<loads> of time (really).
You should understand, that's not enough to just run through changelogs.
It has to be verified that none entry is skipped.
The only reliable (for sake of interface, and, trivially, presence) source of
verification
is C<dpkg-parsechangelog>.
And here's the fork-mare.
C<perl> forks C<shell>, then C<perl>, then C<perl> again.
There seems to be fork of C<tail> too.
And that for B<each> entry.
(Not to count C<gunzip> to decompress the changelog.)
C<loadavg> climbs over 1.50..2.00
You've got the picture.
Although that's where choice begins.

C<v0.2.12>
That happens that C<urgency=high>, probably when it's that really high, is
expressed in blocks (like this: C<urgency=HIGH>).
B<$RE{d}{changelog}> is case-keeping, and then B<dpkg-parsechangelog(1)>
strikes back and lowercase.
From now on such manipulations won't fail a particular entry.

=over

=item I<changelog>

C<v0.2.9>
That defaults to C<changelog=5>.
See below.

=item I<changelog=package>

Only one changelog will be checked.
The one that C<eq>s.
The package name is picked from directory name.

=item I<changelog=a>

Only those changelogs will be checked that C<m/^a/>.

=item I<changelog=5>

C<v0.2.9>
That will check all changelogs, although it will look no more than requested
number of entries deep.
(C<v0.2.9>: ~15min for ~1200 changelogs;
C<v0.2.12>: ~30min for ~1300 changelogs and ~6000 subchecks.)
And that has a perfect sense.
Do you know that C<cron> once changed it's name to C<Cron>
(beware leading block) (cron_3.0pl1-46)?
C'mon, it has happened 12 (tweleve) years ago!
(And you know what?
That default is pretty fair (liblog-log4perl-perl_1.16-1).
Probably it should look for time passed but entry number.)

=item I<changelog=-5>

C<v0.2.9>
That's different.
It will check as many entries as possible (there are changelogs what
B<$RE{d}{changelog}> finds out more entries than B<dpkg-parsechangelog>
(C<dpkg_1.2.13> vs C<dpkg_0.93.79>),
but if the offending record is more than that far from top then it's reported
and otherwise ignored.
(C<v0.2.9>: ~3h for ~1200 packages, ~45000 successful subchecks;
C<v0.2.12>: ~5¼h for ~1300 changelogs, ~63000 successful subchecks.)

=item I<changelog=_5>

C<v0.2.12>
That's a mix of I<changelog=5> and I<changelog=-5>
(thanks to irda-utils_0.9.18-8.1 and mime-support_3.49-1).
It goes no more than configured entries deep and ignores (and reports) any
errors.

=item I<changelog=0>

B<(bug)>
C<v0.2.9>
That will check all changelogs, check all possible entries and B<BAIL_OUT> off
first failure.
Shortly -- don't.
You're warned.
(Although, do it.
F<t/changelog.t> will give up pretty soon.)

=back

To slightly sweeten all that, F<t/changelog.t> attempts to filter
duplicates.
And it B<BAIL_OUT>s upon first failure.

=item B<package> of F<t/package.t>

C<v0.2.10>
Nothing special.
Output of C<dpkg-query -f '${Package}\n' -W> is matched against
B<m/^$RE{debian}{package}$/>.
Probably should parse F<*_Packagees>.

=item B<source> of F<t/archive.source.t>

C<v0.2.3>
Inspects all records in F</var/lib/apt/lists/*_Sources>, extracts I<Files:>
entries, then collects trailing filenames.
They are matched against
C<m/^$RE{debian}{archive}{source_1_0}$/>,
C<m/^$RE{debian}{archive}{patch_1_0}$/>,
C<m/^$RE{debian}{archive}{source_3_0_native}$/>,
C<m/^$RE{debian}{archive}{source_3_0_quilt}$/>,
C<m/^$RE{debian}{archive}{patch_3_0_quilt}$/>,
and
C<m/^$RE{debian}{archive}{dsc}$/>
(in fact C<||>).
If none matches then it will be reported at the end.
C<m/$RE{debian}{archive}{changes}/> is missing here because there is
no source of such on no-build system.

=item B<version> of F<t/version.t>

C<v0.2.10>
Again nothing special.
Output of C<dpkg-query -f '${Version}\n' -W> is matched against
B<m/^$RE{debian}{version}$/>.
Probably should parse F<*_Packages> too.

=back

If any test string fails I need to know what and how.
To provide that info I've picked B<Test::Differences>
(maybe there's other option I'm not aware of?)
(I'm, B<Test::Deep>).
C<v0.60> of B<T::D> closes
B<[38320@rt.cpan.org]> and B<[41241@rt.cpan.org]>.

=cut

my $engine = Module::Build->new(
  module_name        =>                q|Regexp::Common::debian|,
  dist_author        =>      q|Eric Pozharski <whynot@cpan.org>|,
  license            =>                                  q|lgpl|,
  sign               =>                                        1,
  requires           =>
  { q|perl|   =>  q|5.10.0|,
    q|version|    =>   0.50,
    q|Regexp::Common| => ''                                    },
  build_requires     =>
  { q|File::Temp|        =>   '',
    q|Test::Differences| => 0.60,
    q|Test::More|        =>   ''                               },
  dynamic_config         =>                                    1,
  configure_requires => {             q|Module::Build| => 0.38 },
  PL_files           =>
  { ( -f q|README.PL|  ? ( q|README.PL|  =>  q|README| ) : ()),
    ( -f q|Changes.PL| ? ( q|Changes.PL| => q|Changes| ) : ()) },
  add_to_cleanup     =>
  [ qw|              *.tmp *.bak skip_* |,
    ( -f q|README.PL|  ? q|README|  : ()),
    ( -f q|Changes.PL| ? q|Changes| : ())                      ] );

my $y_choice = qx|perl -Mt::TestSuite -we 't::TestSuite::RCD_show_y_choice'|;
chomp $y_choice;
$engine->build_requires( ($y_choice || q|YAML::Tiny|) => '' );

$engine->create_build_script();

=head1 AVAILABILITY

Distribution --
L<http://search.cpan.org/dist/Regexp-Common-debian/>

=head1 BUGS

Please report here --
L<http://rt.cpan.org/Public/Dist/Display.html?Name=Regexp-Common-debian>

=head1 COPYRIGHT AND LICENSING

=over

=item *

Copyright 2008--2010, 2014 Eric Pozharski <whynot@cpan.org>

=item *

AS-IS, NO-WARRANTY, HOPE-TO-BE-USEFUL

=item *

GNU Lesser General Public License v3
L<http://www.gnu.org/licenses/lgpl.txt>

=back
